File: functions/src/setSharedCode.ts

import {onDocumentWritten} from 'firebase-functions/v2/firestore';
import {getFirestore} from 'firebase-admin/firestore';
import * as logger from 'firebase-functions/logger';

import * as firestoreUtils from './firestoreUtils';
import * as stringUtils from './stringUtils';


export const setShareCodes = onDocumentWritten('/lists/{listId}', async (event: any): Promise<void> => {
  const before = event.data.before.data();
  const apre = event.data.after.data();
  const listId = event.params.listId;

  const didSetToShared: boolean =
    apre.shared == true && before.shared != apre.shared &&
    (apre.shareCodeForViewer == undefined || apre.shareCodeForViewer == '') &&
     (apre.shareCodeForEditor == undefined || apre.shareCodeForEditor == '')
    ;

  if (!didSetToShared) {
    return;
  }

  const listPath = `/lists/${listId}`;
  const docRef = getFirestore().doc(listPath);
  const listInfo = (await docRef.get()).data();
  if (listInfo == undefined) {
    logger.log('Could not get data from list');
    return;
  }

  const userId = listInfo.ownerId;

  const shareCodes = await getUnusedShareCodes();
  const viewerCode = shareCodes[0];
  const editorCode = shareCodes[1];
  logger.log('viewerCode', viewerCode);
  logger.log('editorCode', editorCode);

  const codes = {shareCodeForViewer: viewerCode, shareCodeForEditor: editorCode};
  await docRef.set(codes, {merge: true});

  const userPath = `users/${userId}`;
  const userDocRef = getFirestore().doc(userPath);
  const userInfo = (await userDocRef.get()).data();

  const data = {
    ownerUserId: userId,
    ownerListId: listId,
    ownerName: userInfo?.name ?? 'Unknown user',
    listName: listInfo.name,
    listType: listInfo.listType,
  };

  let sharedListPath = `sharedLists/${viewerCode}`;
  logger.log('sharedListPath', sharedListPath);
  let sharedListDocRef = getFirestore().doc(sharedListPath);
  await sharedListDocRef.set({...data, shareType: 'viewer'}, {merge: true});

  sharedListPath = `sharedLists/${editorCode}`;
  sharedListDocRef = getFirestore().doc(sharedListPath);
  await sharedListDocRef.set({...data, shareType: 'editor'}, {merge: true});
});

export async function getUnusedShareCodes(): Promise<string[]> {
  const shareCodes = [];
  let shareCode = '';
  let alreadyExists = true;

  for (let i=0; i<2; i++) {
    while (alreadyExists) {
      shareCode = stringUtils.getRandomString(8);
      alreadyExists = await firestoreUtils.getCodeExists(shareCode);
      // console.log(shareCode, alreadyExists, i);
    }
    shareCodes.push(shareCode);
    alreadyExists = true;
  }

  return shareCodes;
}

File: functions/set_shared_code.py
from firebase_functions_v2.firestore import on_document_written
from firebase_admin import firestore
import firebase_functions.logger as logger
from . import firestoreUtils
from . import stringUtils

async def set_share_codes(event):
    before = event.data.before.data()
    apre = event.data.after.data()
    listId = event.params['listId']

    didSetToShared = apre.get('shared') == True and before.get('shared') != apre.get('shared') and \
                     (apre.get('shareCodeForViewer') == None or apre.get('shareCodeForViewer') == '') and \
                     (apre.get('shareCodeForEditor') == None or apre.get('shareCodeForEditor') == '')

    if not didSetToShared:
        return

    listPath = f"/lists/{listId}"
    doc_ref = firestore.client().document(listPath)
    list_info = (await doc_ref.get()).to_dict()
    if list_info == None:
        logger.log('Could not get data from list')
        return

    userId = list_info['ownerId']

    share_codes = await get_unused_share_codes()
    viewer_code = share_codes[0]
    editor_code = share_codes[1]
    logger.log('viewerCode', viewer_code)
    logger.log('editorCode', editor_code)

    codes = {'shareCodeForViewer': viewer_code, 'shareCodeForEditor': editor_code}
    await doc_ref.set(codes, merge=True)

    user_path = f"users/{userId}"
    user_doc_ref = firestore.client().document(user_path)
    user_info = (await user_doc_ref.get()).to_dict()

    data = {
        'ownerUserId': userId,
        'ownerListId': listId,
        'ownerName': user_info.get('name', 'Unknown user'),
        'listName': list_info['name'],
        'listType': list_info['listType']
    }

    shared_list_path = f"sharedLists/{viewer_code}"
    logger.log('sharedListPath', shared_list_path)
    shared_list_doc_ref = firestore.client().document(shared_list_path)
    await shared_list_doc_ref.set({**data, 'shareType': 'viewer'}, merge=True)

    shared_list_path = f"sharedLists/{editor_code}"
    shared_list_doc_ref = firestore.client().document(shared_list_path)
    await shared_list_doc_ref.set({**data, 'shareType': 'editor'}, merge=True)

async def get_unused_share_codes():
    share_codes = []
    share_code = ''
    already_exists = True

    for i in range(2):
        while already_exists:
            share_code = stringUtils.get_random_string(8)
            already_exists = await firestoreUtils.get_code_exists(share_code)
            # print(share_code, already_exists, i)
        share_codes.append(share_code)
        already_exists = True

    return share_codes
